osu.Server.Queues.ScoreStatisticsProcessor/Helpers/WebRequestHelper.cs

```diff
@@ -0,0 +1,222 @@
+                if (token == null)
+                    return null;
+
+                var requestMsg = new HttpRequestMessage(HttpMethod.Get, $"{shared_interop_domain}/api/v2/scores/{score.id}");
+                requestMsg.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);
+                var responseMsg = http.Send(requestMsg);
+
+                if (!responseMsg.IsSuccessStatusCode || responseMsg.Content.ReadFromJsonAsync<ScoreResponse>().Result is not ScoreResponse scoreResponse)
+                {
+                    SentrySdk.CaptureMessage($"Failed to retrieve score rank from API: received status code {responseMsg.StatusCode}, content: {responseMsg.Content.ReadAsStringAsync().Result}",
+                        SentryLevel.Warning);
+                    return null;
+                }
+
+                return scoreResponse.rank_global;
+            }
+            catch (Exception e)
+            {
+                SentrySdk.CaptureException(e, scope => scope.Level = SentryLevel.Warning);
+                return null;
+            }
+        }
+
+        private static string? retrieveAccessToken()
+        {
+            if (string.IsNullOrEmpty(api_client_id) || string.IsNullOrEmpty(api_client_secret))
+            {
+#if !DEBUG
+                throw new InvalidOperationException($"Attempted API call without client id and secret set!");
+#endif
+                return null;
+            }
+
+            if (accessToken != null && accessTokenExpiry != null && DateTimeOffset.Now <= accessTokenExpiry.Value.AddMinutes(-5))
+                return accessToken;
+
+            var requestMsg = new HttpRequestMessage(HttpMethod.Post, $"{shared_interop_domain}/oauth/token");
+            requestMsg.Content = new FormUrlEncodedContent([
+                new KeyValuePair<string, string>("client_id", api_client_id),
+                new KeyValuePair<string, string>("client_secret", api_client_secret),
+                new KeyValuePair<string, string>("grant_type", "client_credentials"),
+                new KeyValuePair<string, string>("scope", "public"),
+            ]);
+
+            var responseMsg = http.Send(requestMsg);
+
+            if (!responseMsg.IsSuccessStatusCode
+                || responseMsg.Content.ReadFromJsonAsync<TokenResponse>().Result is not TokenResponse tokenResponse
+                || string.IsNullOrEmpty(tokenResponse.access_token))
+            {
+                SentrySdk.CaptureMessage($"Failed to retrieve access token for API call: received status code {responseMsg.StatusCode}, content: {responseMsg.Content.ReadAsStringAsync().Result}", SentryLevel.Warning);
+                return null;
+            }
+
+            accessToken = tokenResponse.access_token;
+            accessTokenExpiry = DateTimeOffset.Now.AddSeconds(tokenResponse.expires_in);
+            return accessToken;
+        }
+
+        // ReSharper disable All
+        private class TokenResponse

```

Comment:
Can probably use `osu.Game.Online.API.OAuthToken`. And arguably, the whole `OAuth` class, which would also mean you'd be using refresh tokens correctly?

Reply:
Right.